### DISCUSSION TITLE
Project-level data retention settings

### DISCUSSION THREAD

**Discussion by sudhanshu746** [langfuse/langfuse#2268] [Ideas]
### Describe the feature or potential improvement
Thanks for providing this great tool, a feature could be a retention settings for trace purge automatically like all traces > 30 days old.

**Comment by marcklingen** [comment]
Agree, with the migration to Langfuse v3, setting expiration dates on single rows will become simpler to handle data retention at scale in langfuse. Thanks for raising this!

**Comment by AlithAnar** [comment]
If there is a safe way to delete traces from PostgreSQL in current v2 setup could you please share from which tables can we manually delete records? are they connected with foreign keys with cascading deletetion feature or we have to manually delete some disconnected remnants?

**Comment by msabramo** [comment]
I am also interested in this. Folks in my org would like to be able to limit data storage to 24 hours to comply with regulations.

LangSmith has this feature: https://docs.smith.langchain.com/self_hosting/configuration/ttl

**Comment by marcklingen** [comment]
Quick update on this thread, retention capabilities will be added in the upcoming v3 of Langfuse as a project-level setting.

### Langfuse Cloud

If you run on Langfuse Cloud, please reach out to us and we can manage retention settings for you project on our end internally.

### Self-hosted

If you run self-hosted, you can safely delete data from the following tables. You can easily run these queries on a cron schedule. If you db is huge, please keep in mind that vacuuming might impact db performance while it is happening.

#### v3 (Clickhouse)

```sql
delete from traces
where project_id = ''
and timestamp < now() - interval N days

delete from observations
where project_id = ''
and start_time < now() - interval N days

delete from scores
where project_id = ''
and timestamp < now() - intervval N days 
```

#### v2 (Postgres)

```sql
DELETE FROM traces
WHERE "timestamp" < '2024-08-01'

DELETE FROM observations
WHERE "start_time" < '2024-08-01'

DELETE FROM scores
WHERE "timestamp" < '2024-08-01'
```

### Q&A

> are they connected with foreign keys with cascading deletetion feature or we have to manually delete some disconnected remnants?

No foreign key constraints are applied that would block a deletion of records. Background: Records need to be compliant with data privacy regulation as well as data needs to be ingested into langfuse without a constraint on the order of records.

**Reply by IsraelShok** [reply]
@marcklingen 
Just to understand, in V3, the project-level setting will be for self-hosted also? or just to LangFuse CLoud users?
Thanks

**Reply by marcklingen** [reply]
v3 will enable more programmatic deletes at the infrastructure level. Currently, in Postgres, this can lead to unpredictable behavior due to vacuuming. We will add project-level retention settings once v3 is released.

**Reply by secsilm** [reply]
> v3 will enable more programmatic deletes at the infrastructure level. Currently, in Postgres, this can lead to unpredictable behavior due to vacuuming. We will add project-level retention settings once v3 is released.

Is project-level retention settings available in v3?

**Reply by marcklingen** [reply]
On Langfuse Cloud (yes), we will productionize this to make this available for teams self-hosting as well

**Reply by secsilm** [reply]
> On Langfuse Cloud (yes), we will productionize this to make this available for teams self-hosting as well

Is there a release plan?

**Reply by marcklingen** [reply]
Yes, but we generally do not communicate ETAs as roadmaps can change. We will update this thread once this is released.

**Reply by RogerSik** [reply]
When this setup in Clickhouse (langfuse v3) do we need also to do a retentionPolicy for the s3 bucket? Or will it here automatically deleted when removed from Clickhouse?

**Reply by Steffen911** [reply]
@RogerSik Yes, S3 is separate from ClickHouse and would need its own Lifecycle rule or deletion setup to only retain data for a limited period of time.

**Reply by peterghaddad** [reply]
Hey @marcklingen, we did some testing on Postgres and noticed that the traces, observations, and score tables do not contain any data.

Is this expected? Due to this, i'm not seeing a need to run these TTL commands on Postgres:

```sql
DELETE FROM traces
WHERE "timestamp" < '2024-08-01'

DELETE FROM observations
WHERE "start_time" < '2024-08-01'

DELETE FROM scores
WHERE "timestamp" < '2024-08-01'
```

We are running v1.5.1 helm-chart version of Langfuse.

**Reply by Steffen911** [reply]
@peterghaddad With our V3 release of Langfuse we shifted the trace, observation, and score data to ClickHouse (https://langfuse.com/blog/2024-12-langfuse-v3-infrastructure-evolution). We kept the old tables around without data for backwards compatibility and a non-breaking migration path. 

For most new deployments running those queries on Postgres won't have an effect, yes.

**Reply by akashkaveti** [reply]
Hi @marcklingen, Instead of manual DELETE operations, can we implement ClickHouse TTL on these tables for automatic data cleanup?
```
ALTER TABLE traces MODIFY TTL timestamp + INTERVAL N DAY DELETE;
ALTER TABLE observations MODIFY TTL start_time + INTERVAL N DAY DELETE;
ALTER TABLE scores MODIFY TTL timestamp + INTERVAL N DAY DELETE;

```

**Reply by Steffen911** [reply]
@akashkaveti You can do this manually for self-hosted instances, yes.

**Reply by synv-2692** [reply]
Hi @Steffen911, Can I use ClickHouse’s TTL feature combined with conditions to delete old data
```sql
ALTER TABLE traces
MODIFY TTL
    timestamp + INTERVAL 30 DAY DELETE WHERE project_id = 'project_abc',
    timestamp + INTERVAL 90 DAY DELETE WHERE project_id = 'project_xyz';
```

**Reply by Steffen911** [reply]
@synv-2692 Yes, that should work too for a self-hosted deployment.

**Reply by Steffen911** [reply]
You may want to patch the observations and scores table as well.

**Comment by sudhanshu746** [comment]
Thank you, @marcklingen for sharing options.I am closing this thread for now.

**Comment by nitinkalra-bst** [comment]
are you not creating these tables partitioned by date? can't we just drop the older partitions?

**Reply by nitinkalra-bst** [reply]
@marcklingen please check this. It takes a lot of time to run delete command

**Reply by marcklingen** [reply]
Currently they are not partitioned b date but indexed by date. This will be changed in [v3](https://github.com/orgs/langfuse/discussions/1902) as the data will be more immutable and thus easier to partition.

**Comment by bharath-muthineni** [comment]
@sudhanshu746 , is there any open request or github issue, just to know when this was implemented? :sweat_smile:

**Reply by sudhanshu746** [reply]
@bharath-muthineni, as per Marc, retention policy has been implemented on V3. for Langfuse V2, we are deleting traces with cron jobs  https://github.com/orgs/langfuse/discussions/2268#discussioncomment-10201896 r

**Reply by secsilm** [reply]
@sudhanshu746 retention is still being developped according to Marc: https://github.com/orgs/langfuse/discussions/2268#discussioncomment-11785959

And, can you share your crontab script for retension? I wrote a python script to do this, but I found that it will cost hours, very slow. There are ~300k traces in a day.

**Reply by marcklingen** [reply]
On v3 these deletes are comparatively faster. On v2 foreign keys in postgres can make these slow, we have fixed this.

**Reply by secsilm** [reply]
@marcklingen On v3, should I also use the method provided below?

https://github.com/orgs/langfuse/discussions/2268#discussioncomment-10201896

**Reply by marcklingen** [reply]
@Steffen911 can share queries that work on clickhouse/v3, they are very similar

**Reply by Steffen911** [reply]
Sure. One could use queries like this:
```
# Traces
delete from traces
where project_id = ''
and timestamp < now() - interval N days

# Observations
delete from observations
where project_id = ''
and start_time < now() - interval N days

# Scores
delete from scores
where project_id = ''
and timestamp < now() - intervval N days 
```

**Reply by marcklingen** [reply]
thank you, added this to the [comment](https://github.com/orgs/langfuse/discussions/2268#discussioncomment-10201896) above for discoverability

**Reply by TheErk** [reply]
Wouldn't it be possible to use clickhouse TTL instead?

```sql
# traces
ALTER TABLE traces
    MODIFY TTL toDate(timestamp) + INTERVAL 30 DAY
# observations
ALTER TABLE observations
    MODIFY TTL toDate(timestamp) + INTERVAL 30 DAY
# scores
ALTER TABLE scores
    MODIFY TTL toDate(timestamp) + INTERVAL 30 DAY
```
etc....

**Reply by Steffen911** [reply]
@TheErk Yes, that would be possible and should work well.

**Reply by Pratish315** [reply]
Hi @marcklingen, are these `ALTER TABLE` statements built into Langfuse or would I need to exec into the pods and run these queries? (for self-hosted)

Is there any plan to build this into Langfuse so the engineer doesn't have to run this? (having a retention parameter in values file within in the helm chart)

**Reply by Steffen911** [reply]
@Pratish315 Those queries would run within ClickHouse.

There is a data retention feature within Langfuse, but it's part of the commercial offering. See https://langfuse.com/docs/data-retention for details.

**Reply by nkalra0123** [reply]
how to manage minio disks in the sekf hosted version? Do you have any suggestions on that? i think they take the most disk

**Reply by Steffen911** [reply]
@nkalra0123 With the Langfuse retention feature we should delete things from MinIO automatically. 
For other tiers, minio lifecycle rules may work: https://min.io/docs/minio/linux/administration/object-management/object-lifecycle-management.html#object-expiration.
